### DISCUSSION TITLE
How to improve AWS Lambda cold start problem while having large schemas ?

### DISCUSSION THREAD

**Discussion by mouhannad-sh** [fastify/fastify#5718] [Q&A]
Hi, 
At work we used to have a monolithic NodeJS Express API running in AWS lambda behind an AWS HTTP API gateway configuration that was performing really well in terms of cold start but as the business grew so did our demand for a stable API for 3rd party integrations.

we are a pretty small team of 3 developers and while Express with Zod was serving us well. we started facing issues scaling the codebase and we found good solutions to our problems in Fastify and its ecosystem. especially the schema and first class typescript support !

Fastify allowed us to ship faster and more importantly it felt a lot safer with a great DX too!

But right now I'm facing issues with the cold start performance of our monolithic fastify API and based on my manual testing and benchmarking it seems like the issue is that our schemas are too specific and rather large and most of the API server cold start time is spent on compiling the AJV + Typebox schema.

Back in the express version. we used to build context specific zod validation schema in-flight rather than during startup. so the performance was pretty acceptable - not the fastest but pretty acceptable - but most importantly our cold start penalty was small.

In attempting to improve the cold start performance, I followed this lovely article by [Manuel Spigolon](https://hashnode.com/@Eomm)
https://backend.cafe/how-to-unlock-the-fastest-fastify-server-startup

At the start, it was pretty helpful and we noticed an improvement in our cold starts but it was still taking around 1000ms - 1600ms for startup. but when we eliminate our schemas, the entire api startup is within 50ms - 100ms which is fast enough for us.

I was hoping for an official support for async schema compilation or partial schema compilation. but I can't seem to find any reference of that anywhere I looked.

it would be perfect - for my case - if somehow there was a recommended methodology or API where we could choose which parts of the schema to compile at runtime. 

our main benefits of using Fastify in case you're wondering are:
1. schema validation for the request and response
2. the openapi documentation that we can easily layer on top with fastify is such a bliss !
3. excellent typescript support
4. the inject api makes testing a lot easier
5. It's FAST ! - when the lambda is warm - 
6. as business we try to integrate with many 3rd party partners. so the features above allows us to ship our own features and change things in our platform and at the same time guarantee a decent documentation that our partners can rely on.

our arm64 with 2048 RAM lambdas are running on nodeJS 20 and making use of `@fastify/aws-lambda` and top level await

Additionally, we are using the official Typebox provider to build our schema

I'm planning to do a deep dive into our internal workflow and code-flow to figure out (a hack) how we can get better cold starts and still achieve the above benefits.

But before I start working on that, I'm wondering if anyone in the community has been able to find a better approach for using fastify apis in aws lambda.

Thanks !

**Comment by Eomm** [comment]
Interesting feature.

I think it is doable by implementing a custom `schemaController` (example: https://github.com/Eomm/joi-compiler ).

The `factory.buildValidator` function could be a lazy loading factory that compiles the schema the first time it is being called with the input `data`.

One edge case to handle: what if the schema is invalid and it does not compile? The route will never work and the server is up&running.

> At the start, it was pretty helpful and we noticed an improvement in our cold starts but it was still taking around 1000ms - 1600ms for startup

Out of curiosity: how was the start time without the pre-compilation?

**Comment by mouhannad-sh** [comment]
Thanks @Eomm 

I've been meaning to reply to you for months but I didn't have the time to experiment enough.

Since my post above, our product grew and we added more schemas than before.

in Sep 2024 our OpenAPI json schema was about ~650kb
in December 2024 it jumped to a whopping ~1.2MB.

As a result our cold start `Billed duration` jumped to 5200ms - 6300ms and our customers noticed it (rightfully so !) 

after our customers complained I managed to justify spending time on optimizations and here's how it went:

### ⚠️ Disclamer
The code below is not battle tested and this hacky optimization is specifically tested  on AWS Lambda. if you're running your code on a different provider or on an actual server. probably none of this is necessary !


### Optimize the Schema size and upgrade to Node 22 runtime 
At first, my thinking was around memory consumption as I saw that a cold start request to a 404 link (should be low footprint) was consuming around 450++ mb
I immediately noticed that we can improve the schema size by utilizing references, so that's exactly what I did !
I managed to quickly shave the size of the schema to ~540kb.

Deployed to AWS and sent a cold start request to a 404 link.... Nothing changed at all ! cold start times and memory consumption didn't budge.

### Try pre-compilation and loading schema from file using `StandaloneValidation` & `StandaloneSerializer` 

Deployed to AWS and sent a cold start request to a 404 link.... Nothing changed at all ! cold start times and memory consumption didn't change .

### Upgrade to Fastify 5.x (from 4.28.1)
Fixed conflicts and update AJV & Typebox to latest

Deployed to AWS and sent a cold start request to a 404 link.... and I noticed a small improvement, cold starts are now roughly between 5000ms to 5500ms. We were planning on the upgrade anyway so that's a win in my book !

### First attempt at fiddling with SchemaController

> I think it is doable by implementing a custom schemaController (example: https://github.com/Eomm/joi-compiler ).
> The factory.buildValidator function could be a lazy loading factory that compiles the schema the first time it is being called with the input data.

To be honest the documentation around the validator and schema factories aren't great. and it took me a good amount of trial and error until I managed to write my own validator factory.

Deployed to AWS and sent a cold start request to a 404 link.... and I noticed a marginal improvement... cold starts are now around 4500ms to 4700ms.

At this point I was very suspicious as I have built larger apis with larger validation schemas in the past with express and cold start was never of a real issue....

Another thing I noticed. once the lambda is warm. I can get responses from our api in under 30ms !

##### Wrap all [Typebox](https://github.com/sinclairzx81/typebox) schemas in arrow functions 🚀 
At this point I decided to go back to the basics of Javascript. Looking at our code. All schemas are defined like this 

```ts
const schema = {
    tags: ["Jobs"],
    params: t.Object({ idRef: t.String({ description: "id or job reference", minLength: 1 }) }),
    description: "Get specific Job by id or Ref",
    response: {
      200: t.Object({ data: API_Job_Ref}),
      404: API_NotFound_Ref,
    },
  }
```
So what jumped to me here is that we are executing a lot of Typebox functions during startup.

We have at the moment 102 routes each with a schema and a purpose.

So I removed all routes and kept our biggest schema route whos' schema references multiple other models

Then I deployed it and the cold start time improved by a few 100s of ms, it was still averaging around 4000ms.

So I just wrapped the schema with a `() => schema` and immediately noticed that the `schemaController` `validator` stopped working and that route stopped validating the input. I'm guessing that's because Fastify expects the `schema` to be an object instead of a function.

So I added a `preValidation` hook  with the following quick and dirty code
```ts
app.addHook("preValidation", async (req, res) => {
    Object.values(req.server.getSchemas()).forEach((schema) => {
      if (ajv.schemas[schema.$id]) return;
      ajv.addSchema(schema as AnySchema);
    });
    const _schema = req.routeOptions.schema as FastifySchema | (() => FastifySchema); ;
    if (_schema) {
      const schema = typeof _schema === "function" ? _schema() : _schema;
      if (schema.body) {
        const validate = ajv.compile(schema.body);
        if (!validate(req.body) && validate.errors) {
          throw new APIBadRequestError(
            `body${validate.errors?.map((e) => e.instancePath + " " + e.message).join(", ")}`,
          );
        }
      }
      if (schema.params) {
        const validate = ajv.compile(schema.params);
        if (!validate(req.params) && validate.errors) {
          throw new APIBadRequestError(
            `url path${validate.errors?.map((e) => e.instancePath + " " + e.message).join(", ")}`,
          );
        }
      }
      if (schema.querystring) {
        const validate = ajv.compile(schema.querystring);
        if (!validate(req.query) && validate.errors) {
          throw new APIBadRequestError(
            `searchParams${validate.errors?.map((e) => e.instancePath + " " + e.message).join(", ")}`,
          );
        }
      }
      if (schema.headers) {
        const validate = ajv.compile(schema.headers);
        if (!validate(req.headers) && validate.errors) {
          throw new APIBadRequestError(
            `headers${validate.errors?.map((e) => e.instancePath + " " + e.message).join(", ")}`,
          );
        }
      }
    }
  });
```
Now the api can validate - missing response validation and schema validation tho ! -

Deployed to AWS and 💥  Cold start time is at ~600ms 🚀 

After that I brought back all the routes and the schemas but made sure to refactor all of them into functions and removed my `schemaController` factories

Deployed to lambda ... cold start is at around ~900ms !


<img width="246" alt="image" src="https://github.com/user-attachments/assets/818dcb34-5a5b-481a-8b5c-77a1fc65b249" />


### Fixing the OAPI schema json generation to be "acceptable for our usecase"

We actually rely on the schema generated by fastify to do a few things: 
- Partners documentation website site 
- We generate our client side TS SDK library with [openapi-ts](https://github.com/openapi-ts/openapi-typescript). So we can't compromise on the schema.
- I hook it up with my work GPT 🌚 

So I created an endpoint for it and stiched the OAPI spec dynamically by inspecting every route in the server and the schemas added via `fastify.addSchema()`

###  Validating the schema
My solution thus far is not perfect and has many gaps but for us "it works"...
During development we use Vite to run the dev server. so we also build the schema every time the developer makes a change and if the schema is invalid for any reason. the devServer will throw an error forcing the dev to fix it.

For production. building the schema is a pre-check after we build our nodejs bundle just before we deploy to production / staging

## Conclusion

I wish if Fastify could accept `schema` as a function instead of an object and to have a global option to defer building the schema to a later stage of some sort.

Regarding the solution above: there's a lot of implementation details behind the scenes there (e.g. inspecting fastify routes reliably) and much of my current is solution is not battle tested at all. I just tinkered until I managed to get it to "work" for our usecase.

I'm pretty sure there's a lot of room for improvement but for me this is a good start !

**Reply by Uzlopak** [reply]
> I wish if Fastify could accept schema as a function instead of an object and to have a global option to defer building the schema to a later stage of some sort.

Hmm. Yeah somehow... what if a schema is identical to another schema? There could be some kind of improvement there...

**Reply by mouhannad-sh** [reply]
> Hmm. Yeah somehow... what if a schema is identical to another schema? There could be some kind of improvement there...

That's a valid point.

During development, I would say this is an easy thing to handle as fastify/ajv would probably throw an exception in case of an id collision.

If there's no collision but it's just compiling  an identical schema that was compiled before... maybe a cache key of or a reference of some sort can help with this ?

**Reply by Uzlopak** [reply]
ajv solves this afaik with a WeakMap. Maybe we should also put the objects into a WeakMap.

**Reply by mouhannad-sh** [reply]
> ajv solves this afaik with a WeakMap. Maybe we should also put the objects into a WeakMap.

I haven't tried that yet but I feel the problem is more with Typebox schemas being on the top level + AJV compiling everything at startup.
